hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java

```diff
@@ -1130,6 +1134,21 @@ public void close() {
     evictStateMachineCache();
     executor.shutdown();
     metrics.unRegister();
+
+    // if datanodeService is stopped , it indicates this `close` originates
+    // from `HddsDatanodeService.stop()`, otherwise, it indicates this `close` originates
+    // from ratis.
+    if (datanodeService != null && !datanodeService.isStopped()) {
+      LOG.error("Container statemachine is closed by ratis, terminating HddsDatanodeService");
+      // wait a while for other pipeline's ContainerStateMachine.close() called.
+      try {
+        Thread.sleep(10000);
```

Comment:
Is there a more reliable way to wait for other pipeline closure other than sleep here?

And what happens if there are still unclosed pipeline after 10 seconds' wait?

Reply:
DN has the capability to recover from a immediate crash or kill(because of OOM).  So 10s here is try to let other pipelines persist as many as possible state to disk. Shutdown immediately, wait 5s or 10s, has no big difference. 10s is just an experience value.